### ISSUE TITLE
Wrong error is raised for property of nn.Module (again)

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#49726]
## 🐛 Bug

Same as https://github.com/pytorch/pytorch/issues/33934. See there for details. The same issue exists in 1.7.0. 

```py
In [7]: class A(torch.nn.Module):
   ...:     @property
   ...:     def attr(self): return self.c
   ...:
   ...:

In [8]: A().attr
---------------------------------------------------------------------------
ModuleAttributeError                      Traceback (most recent call last)
<ipython-input-8-d29dd47872e1> in <module>
----> 1 A().attr

~/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py in __getattr__(self, name)
    777                 return modules[name]
    778         raise ModuleAttributeError("'{}' object has no attribute '{}'".format(
--> 779             type(self).__name__, name))
    780
    781     def __setattr__(self, name: str, value: Union[Tensor, 'Module']) -> None:

ModuleAttributeError: 'A' object has no attribute 'attr'
```

It is totally unfixed because the PR that claims to have fixed it (https://github.com/pytorch/pytorch/pull/34324) uses wrong unit tests (this link points to 1.7.0 branch):
https://github.com/pytorch/pytorch/blob/e85d494707b835c12165976b8442af54b9afcb26/test/test_nn.py#L1065-L1077
The `assertIn`s wouldn't run at all. 

cc @albanD @mruberry @jbschlosser @mpariente @ezyang

**Comment by jonykarki** [comment]
I changed the ModuleAttributeError to derive from Exception instead of AttributeError. It seems like once we're in getattr and the attribute that is not found is one that was inside a function then getattr is called once again to find the outer function this time. Since, the __getattribute__ isn't called the second time (probably because AttributeError was raised for the first check), the outer function isn't found and the ModuleAttributeError is raised on the outer function.
So, deriving from Exception will cause it to stop once the attribute is not found on the first check itself and raises error on the first one. I don't know if my explanation will make any sense but doing what I did seems to work at least for this case.
I haven't changed or written the tests yet.

**Comment by jbschlosser** [comment]
@mruberry and I looked at this. Summary for those who don't want to dig into the original issue / PRs:

As mentioned elsewhere, this is a common problem that occurs when mixing `@property` and `__getattr__` (see [here](https://stackoverflow.com/questions/25115366/property-and-getattr-compatibility-issue-with-attributeerror), [here](http://blog.devork.be/2011/06/using-getattr-and-property_17.html), [here](https://medium.com/@ceshine/python-debugging-pitfall-mixed-use-of-property-and-getattr-f89e0ede13f1), etc.). Common advice is to "not throw AttributeErrors from `@property` methods" (unfortunately, this is easier said than done since the AttributeError can result from a typo, as demonstrated in the original issue). Other [advice](https://stackoverflow.com/questions/11116896/python-why-is-getattr-catching-attributeerrors) simply discourages mixing `@property` and `__getattr__` at all.

This is a _Python_ issue and only results in PyTorch because we define `__getattr__` on `nn.Module`. The question is: "should we provide a workaround for this Python issue?"
  * Possible workaround 1: @jonykarki's PR seems to fix the issue by modifying `__getattr__` to call `__getattribute__` at the end instead of raising an `AttributeError`. However, this depends on some idiosyncrasies in python's handling of `AttributeError`; it somehow avoids infinite recursion from `__getattribute__` calling `__getattr__`, which calls back to `__getattribute__`, etc. It may be safe to depend on this, but it is not straightforward and adds extra call overhead (although this overhead is only incurred in error cases, so it may not matter).
  * Possible workaround 2: Override `__getattribute__` in `nn.Module` instead of implementing `__getattr__`. This was suggested [here](https://stackoverflow.com/questions/25115366/property-and-getattr-compatibility-issue-with-attributeerror) and in @mpariente's original PR. The logic is clear, but the full impact of switching to `__getattribute__` is unclear. @jonykarki has correctly pointed out that several tests explicitly depend on `__getattr__` existing, and @mruberry has brought up the possibility of JIT breakage from this change. More research is needed.
  * Other possible workaround?? (maybe there's another simple fix with minimal impact)

Alternatively, since this is a Python issue, it can be argued that it's not our responsibility to implement a workaround (i.e. point to the advice of "don't mix `@property` and `__getattr__`"). In this case, @jonykarki's PR could be repurposed to simply clean up the unsuccessful `ModuleAttributeError` fix. In the future, we could revisit the possibility of moving from `__getattr__` to `__getattribute__` as a separate issue.

@SsnL : What are your thoughts on whether we should provide a workaround for this?

**Comment by SsnL** [comment]
@jbschlosser I agree that this is in essence a Python issue. However, I still think that we should provide a workaround because this confusing phenomenon is really caused jointly by user and PyTorch (which defines custom `__getattr__`) and the users should not be expected to know this implementation detail, and thus may likely be confused (for instance, they wouldn't even know that they need to search `"property and custom __getattr__"` on google).

**Comment by jbschlosser** [comment]
> @jbschlosser I agree that this is in essence a Python issue. However, I still think that we should provide a workaround because this confusing phenomenon is really caused jointly by user and PyTorch (which defines custom `__getattr__`) and the users should not be expected to know this implementation detail, and thus may likely be confused (for instance, they wouldn't even know that they need to search `"property and custom __getattr__"` on google).

@SsnL I 100% agree with you here! A smooth UX wouldn't require the user to know this implementation detail to properly debug their own code.

My personal opinion is that we should provide a workaround, and it should be the "right" workaround. I think moving the logic from `__getattr__` into `__getattribute__` is the "right" way to go. It is understandable, doesn't depend on specific python idiosyncrasies, and has been adopted in at least one other open-source project (see [Django](https://github.com/encode/django-rest-framework/pull/2530/files#diff-6dc9b019ec1d96c56e0e31dac3bba51cL488)).

Unfortunately, at least one other component of PyTorch depends on the implementation detail of `__getattr__` being defined. Specifically, `torch.fx` swaps out `__getattr__` with a custom version for in `symbolic_trace` for tracing access of parameters. This is why the `__getattribute__` workaround causes the fx tests to fail.

Given this, I think we need more buy-in from other teams before we move forward with the workaround. I suggest that we address reverting the broken `ModuleAttributeError` fix in @jonykarki's PR and open a new issue for the possibility of moving from `__getattr__` to `__getattribute__`, where we can cc / bring in other affected teams.

**Comment by jbschlosser** [comment]
Just opened #50292 to get broader feedback.

@jonykarki How do you feel about your PR removing the broken `ModuleAttributeError` fix and related tests for now?

**Comment by jonykarki** [comment]
I don't know if I can still work on the same PR since I messed it up. I can do it in a new PR and just remove the ModuleAttributeError and fix the related tests if you want.

**Comment by jbschlosser** [comment]
> I don't know if I can still work on the same PR since I messed it up. I can do it in a new PR and just remove the ModuleAttributeError and fix the related tests if you want.

Sure, if you're up for it, a new PR would be awesome!

**Comment by jbschlosser** [comment]
Reopening because the merged PR just reverts the broken old fix.

**Comment by EricCousineau-TRI** [comment]
Ran into this as well, confirmed that this is still an issue on `torch==1.8.1` (diff. case, but readapted to Tongzhou's repro code):
https://github.com/EricCousineau-TRI/repro/blob/779c05c42a8/bug/torch_nested_module_prop_err/README.md

**Comment by EricCousineau-TRI** [comment]
Dunno if this is more effort than rearchitecting to `__getattribute__`, but perhaps it can be "lightly" addressed, e.g. have a FAQ entry:
```md
If you need to use `@property` in `nn.Module`, please note that it may interfere with how `__getattr__` works due to a fundamental limitation in Python,  and it may produce a misleading error message (<link to StackOverflow>). If you decide that you still want `@property`, please consider intercepting exceptions and hoisting the type so you get useful error messages.
```

Extra credit for seeing if it's possible to print this warning in the narrowest set of circumstances? (e.g. a nested `ModuleAttributeError` is received, and use `warnings.warn()` to print link to FAQ?)

It wouldn't capture Tongzhou's issue, but it may've caught my issue where it was recursive within `Module.__getattr__`:
https://github.com/EricCousineau-TRI/repro/blob/75e13f1b1c3e0ace4b365174c8aeb6e65ea77fba/bug/torch_nested_module_prop_err/repro.py
